test/GPv2Encoding.test.ts

```diff
@@ -0,0 +1,211 @@
+      ).to.be.revertedWith("malformed order data");
+    });
+
+    it("should revert if order bytes are too long.", async () => {
+      await expect(
+        encoding.decodeSignedOrdersTest(
+          [],
+          1,
+          ethers.utils.hexlify([...Array(205)].map(() => 42)),
+        ),
+      ).to.be.revertedWith("malformed order data");
+    });
+
+    const sampleOrder = {
+      sellToken: fillBytes(20, 0x01),
+      buyToken: fillBytes(20, 0x02),
+      sellAmount: ethers.utils.parseEther("42"),
+      buyAmount: ethers.utils.parseEther("13.37"),
+      validTo: 0xffffffff,
+      nonce: 0,
+      tip: ethers.constants.WeiPerEther,
+      kind: OrderKind.SELL,
+      partiallyFillable: false,
+    };
+
+    it("should revert for invalid order signatures", async () => {
+      const encoder = new OrderEncoder(domainSeparator);
+      encoder.signEncodeOrder(traders[0], sampleOrder, sampleOrder.sellAmount);
+
+      // NOTE: `v` must be either `27` or `28`, so just set it to something else
+      // to generate an invalid signature.
+      const encodedOrderBytes = ethers.utils.arrayify(encoder.encodedOrders);
+      encodedOrderBytes[139] = 42;
+
+      await expect(
+        encoding.decodeSignedOrdersTest(
+          encoder.tokens,
+          encoder.orderCount,
+          encodedOrderBytes,
+        ),
+      ).to.be.revertedWith("invalid signature");
+    });
+
+    it("should revert for invalid sell token indices", async () => {
+      const lastToken = fillBytes(20, 0x03);
+
+      const encoder = new OrderEncoder(domainSeparator);
+      encoder.signEncodeOrder(traders[0], sampleOrder, sampleOrder.sellAmount);
+      encoder.signEncodeOrder(
+        traders[1],
+        {
+          ...sampleOrder,
+          sellToken: lastToken,
+        },
+        sampleOrder.sellAmount,
+      );
+
+      // NOTE: Remove the last sell token (0x0303...0303).
+      const tokens = encoder.tokens;
+      expect(tokens.pop()).to.equal(lastToken);
+      await expect(
+        encoding.decodeSignedOrdersTest(
+          tokens,
+          encoder.orderCount,
+          encoder.encodedOrders,
+        ),
+      ).to.be.reverted;
+    });
+
+    it("should revert for invalid sell token indices", async () => {
+      const lastToken = fillBytes(20, 0x03);
+
+      const encoder = new OrderEncoder(domainSeparator);
+      encoder.signEncodeOrder(traders[0], sampleOrder, sampleOrder.sellAmount);
+      encoder.signEncodeOrder(
+        traders[1],
+        {
+          ...sampleOrder,
+          buyToken: lastToken,
+        },
+        sampleOrder.sellAmount,
+      );
+
+      // NOTE: Remove the last buy token (0x0303...0303).
+      const tokens = encoder.tokens;
+      expect(tokens.pop()).to.equal(lastToken);
+      await expect(
+        encoding.decodeSignedOrdersTest(
+          tokens,
+          encoder.orderCount,
+          encoder.encodedOrders,
+        ),
+      ).to.be.reverted;
+    });
+  });
+});
```

Comment:
nit: for full coverage you might need a test for `malformed order data` as well.

Reply:
There should be test cases for that as well: https://github.com/gnosis/oba-contracts/blob/1ed83dcb267d73a3fe20c51678814495e6e16621/test/GPv2Encoding.test.ts#L116